src/shared/hibernate-util.c

```diff
@@ -28,441 +29,426 @@
- *  entrusted with full system memory access (for example via /sys/power/resume) or that isn't an already
- *  active swap area!
- *  Otherwise various security attacks might become possible, for example an attacker could silently attach
- *  such a device and circumvent full disk encryption when it would be automatically used for hibernation.
- *  Also, having a swap area on top of encryption is not per se enough to protect from all such attacks.
- *
- * Returns:
- *  1 - Values are set in /sys/power/resume and /sys/power/resume_offset.
- *      ret_hibernate_location will represent matching /proc/swap entry if identified or NULL if not.
- *
- *  0 - No values are set in /sys/power/resume and /sys/power/resume_offset.
-        ret_hibernate_location will represent the highest priority swap with most remaining space discovered in /proc/swaps.
- *
- *  Negative value in the case of error.
- */
-int find_hibernate_location(HibernateLocation **ret_hibernate_location) {
-        _cleanup_fclose_ FILE *f = NULL;
-        _cleanup_(hibernate_location_freep) HibernateLocation *hibernate_location = NULL;
-        dev_t sys_resume = 0; /* Unnecessary initialization to appease gcc */
-        uint64_t sys_offset = 0;
-        bool resume_match = false;
+static void swap_entries_done(SwapEntries *entries) {
+        assert(entries);
+
+        FOREACH_ARRAY(i, entries->swaps, entries->n_swaps)
+                swap_entry_done(i);
+
+        free(entries->swaps);
+}
+
+static int swap_entry_get_resume_config(SwapEntry *swap) {
+        _cleanup_close_ int fd = -EBADF;
+        uint64_t offset_raw;
+        struct stat st;
         int r;
 
-        /* read the /sys/power/resume & /sys/power/resume_offset values */
-        r = read_resume_files(&sys_resume, &sys_offset);
+        assert(swap);
+        assert(swap->path);
+
+        fd = open(swap->path, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
+        if (fd < 0)
+                return -errno;
+
+        if (fstat(fd, &st) < 0)
+                return -errno;
+
+        if (!swap->swapfile) {
+                if (!S_ISBLK(st.st_mode))
+                        return -ENOTBLK;
+
+                swap->devno = st.st_rdev;
+                swap->offset = 0;
+                return 0;
+        }
+
+        r = stat_verify_regular(&st);
         if (r < 0)
                 return r;
 
-        f = fopen("/proc/swaps", "re");
-        if (!f) {
-                log_debug_errno(errno, "Failed to open /proc/swaps: %m");
-                return errno == ENOENT ? -EOPNOTSUPP : -errno; /* Convert swap not supported to a recognizable error */
+        r = get_block_device_fd(fd, &swap->devno);
+        if (r < 0)
+                return r;
+
+        r = fd_is_fs_type(fd, BTRFS_SUPER_MAGIC);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to check if swap file '%s' is on Btrfs: %m", swap->path);
+        if (r > 0) {
+                r = btrfs_get_file_physical_offset_fd(fd, &offset_raw);
```

Comment:
The usage of btrfs_get_file_physical_offset_fd when checking the configured kernel resume offset now requires CAP_SYS_ADMIN for systemd-sleep hibernate @YHNdnzj.

```
allow systemd_sleep_t self:capability sys_admin;
```
Fedora's SELinux policy only allows systemd_sleep_t to use the following capabilities.
```
allow systemd_sleep_t self:capability { linux_immutable sys_resource };
# systemd-sleep needs to set timer for suspend-then-hibernate
allow systemd_sleep_t self:capability2 wake_alarm;
dontaudit systemd_sleep_t self:capability sys_ptrace;
```
https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L1594-L1597

The first ioctl that fails is [BTRFS_IOC_TREE_SEARCH](https://btrfs.readthedocs.io/en/latest/btrfs-ioctl.html).

Reply:
Hmm, sorry, but I have basically 0 knowledge about SELinux... Could you please open an issue there? I don't there's anything to change in systemd codebase.

Reply:
Yeah this should be fixed in Fedora's selinux policy.

Reply:
I merely wonder if it was possible without CAP_SYS_ADMIN or should fallback to skip the check when lacking permission. I do see, that a lot of other systemd services have sys_admin capability, so will make a PR to fix the SELinux policy.